datadog/dogstatsd/base.py

```diff
@@ -552,9 +552,9 @@ def _should_flush(self, length_to_be_added):
         return False
 
     def _flush_buffer(self):
-        self._send_to_server("\n".join(self.buffer))
+        to_send, self.buffer = self.buffer, []
         self._current_buffer_total_size = 0
-        self.buffer = []
+        self._send_to_server("\n".join(to_send))
```

Comment:
There's still a chance another thread will write to the buffer reference, depending on where it was interrupted.

This improves the timing but does not fix the root cause.